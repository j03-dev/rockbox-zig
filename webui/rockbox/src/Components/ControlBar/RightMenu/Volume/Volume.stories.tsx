import type { Meta, StoryObj } from "@storybook/react";

import Volume from "./Volume";
import { fn } from "@storybook/test";

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: "Components/Volume",
  component: Volume,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    volume: {
      control: {
        type: "range",
        min: 0,
        max: 100,
        step: 1,
      },
    },
    onVolumeChange: {
      action: "volume changed",
    },
  },
} satisfies Meta<typeof Volume>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const Default: Story = {
  args: {
    volume: 70,
    onVolumeChange: fn(),
  },
};
