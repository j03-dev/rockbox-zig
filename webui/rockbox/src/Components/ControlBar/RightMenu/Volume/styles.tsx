import styled from "@emotion/styled";

export const Container = styled.div`
  flex: 1;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export default {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  slider: (t: any) => ({
    color: "rgba(0, 0, 0, 0.682)",
    "& .MuiSlider-track": {
      border: "none",
    },
    "& .MuiSlider-thumb": {
      width: 18,
      height: 18,
      backgroundColor: "#fff",
      "&::before": {
        boxShadow: "0 4px 8px rgba(0,0,0,0.18)",
      },
      "&:hover, &.Mui-focusVisible, &.Mui-active": {
        boxShadow: "none",
      },
    },
    ...t.applyStyles("dark", {
      color: "#fff",
    }),
  }),
};
