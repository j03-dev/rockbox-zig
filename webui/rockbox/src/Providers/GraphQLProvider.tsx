import {
  ApolloClient,
  createHttpLink,
  InMemoryCache,
  ApolloProvider,
  split,
  from,
} from "@apollo/client";
import { removeTypenameFromVariables } from "@apollo/client/link/remove-typename";
import { WebSocketLink } from "@apollo/client/link/ws";
import { getMainDefinition } from "@apollo/client/utilities";
import { SubscriptionClient } from "subscriptions-transport-ws";
import { FC, ReactNode } from "react";

const uri =
  process.env.NODE_ENV === "development"
    ? import.meta.env.VITE_APP_API_URL || "http://localhost:6062/graphql"
    : `${origin}/graphql`;

const removeTypenameLink = removeTypenameFromVariables();

const httpLink = createHttpLink({
  uri,
});

const wsLink = new WebSocketLink(
  new SubscriptionClient(uri.replace("http", "ws"))
);

const splitLink = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === "OperationDefinition" &&
      definition.operation === "subscription"
    );
  },
  wsLink,
  from([removeTypenameLink, httpLink])
);
const link = splitLink;

const client = new ApolloClient({
  link,
  cache: new InMemoryCache(),
});

const GraphQLProvider: FC<{
  children: ReactNode;
}> = ({ children }) => {
  return <ApolloProvider client={client}>{children}</ApolloProvider>;
};

export default GraphQLProvider;
