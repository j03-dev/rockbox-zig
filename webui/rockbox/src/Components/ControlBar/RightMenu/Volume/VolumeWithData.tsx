import {
  useAdjustVolumeMutation,
  useGetGlobalSettingsQuery,
} from "../../../../Hooks/GraphQL";
import Volume from "./Volume";
import { FC, useMemo } from "react";

const VolumeWithData: FC = () => {
  const { data, refetch } = useGetGlobalSettingsQuery();
  const [adjustVolume] = useAdjustVolumeMutation();
  const volume = useMemo(() => {
    return Math.min((data?.globalSettings.volume || 0) + 80, 80);
  }, [data]);

  const onVolumeChange = async (newVolume: number) => {
    const steps = Math.min(newVolume, 80) - Math.min(volume, 80);
    await adjustVolume({
      variables: {
        steps,
      },
    });
    await refetch();
  };

  return <Volume volume={volume} onVolumeChange={onVolumeChange} />;
};

export default VolumeWithData;
