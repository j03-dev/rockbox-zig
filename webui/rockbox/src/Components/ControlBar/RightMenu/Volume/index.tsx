import Volume from "./VolumeWithData";

export default Volume;
