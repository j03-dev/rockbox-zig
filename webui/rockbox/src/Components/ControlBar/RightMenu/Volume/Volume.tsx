import { FC, SyntheticEvent, useEffect, useState } from "react";
import Slider from "@mui/material/Slider";
import styles, { Container } from "./styles";

export type VolumeProps = {
  volume: number;
  onVolumeChange: (volume: number) => Promise<void>;
};

const Volume: FC<VolumeProps> = (props) => {
  const [volume, setVolume] = useState<number>(props.volume);

  useEffect(() => {
    setVolume(props.volume);
  }, [props.volume]);

  const handleVolumeChange = async (
    _event: Event | SyntheticEvent<Element, Event>,
    value: number | number[]
  ) => {
    await props.onVolumeChange(value as number);
  };

  return (
    <Container>
      <div style={{ width: 100, marginTop: 5 }}>
        <Slider
          aria-label="Volume"
          value={volume}
          onChange={(_event, value) => setVolume(value as number)}
          onChangeCommitted={handleVolumeChange}
          sx={styles.slider}
        />
      </div>
    </Container>
  );
};

export default Volume;
