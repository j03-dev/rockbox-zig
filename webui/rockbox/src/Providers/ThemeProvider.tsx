import { createContext, useState, FC } from "react";
import { ThemeProvider as EmotionThemeProvider } from "@emotion/react";
import {
  BaseUIDarkTheme,
  BaseUILightTheme,
  DarkTheme,
  LightTheme,
} from "../Theme";
import { BaseProvider } from "baseui";
import { ThemeProvider as MaterialThemeProvider } from "@mui/material/styles";
import { createTheme } from "@mui/material/styles";

export type Theme = "light" | "dark";

export type ThemeContextType = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
};

export const ThemeContext = createContext<ThemeContextType>({
  theme: "light",
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  setTheme: (_theme: Theme) => {},
});

const muitheme = createTheme({
  cssVariables: true,
  palette: {
    primary: {
      main: "#fe099c",
    },
  },
});

export type ThemeProviderProps = {
  children: React.ReactNode;
};

const ThemeProvider: FC<ThemeProviderProps> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>("light");
  return (
    <MaterialThemeProvider theme={muitheme}>
      <ThemeContext.Provider value={{ theme, setTheme }}>
        <EmotionThemeProvider theme={theme === "dark" ? DarkTheme : LightTheme}>
          <BaseProvider
            theme={theme === "dark" ? BaseUIDarkTheme : BaseUILightTheme}
          >
            {children}
          </BaseProvider>
        </EmotionThemeProvider>
      </ThemeContext.Provider>
    </MaterialThemeProvider>
  );
};

export default ThemeProvider;
