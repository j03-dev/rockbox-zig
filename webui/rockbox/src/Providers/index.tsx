import { FC } from "react";
import ThemeProvider from "./ThemeProvider";
import { Provider as StyletronProvider } from "styletron-react";
import { Client as Styletron } from "styletron-engine-atomic";
import { RecoilRoot } from "recoil";

const engine = new Styletron();

export type ProvidersProps = {
  children: React.ReactNode;
};

const Providers: FC<ProvidersProps> = ({ children }) => {
  return (
    <RecoilRoot>
      <StyletronProvider value={engine}>
        <ThemeProvider>{children}</ThemeProvider>
      </StyletronProvider>
    </RecoilRoot>
  );
};

export default Providers;
