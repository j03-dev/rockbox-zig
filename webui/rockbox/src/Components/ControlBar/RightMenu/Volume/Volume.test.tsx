import { render } from "@testing-library/react";
import { vi } from "vitest";
import Volume from "./Volume";
import Providers from "../../../../Providers";

describe("Volume", () => {
  it("should render", () => {
    const { container } = render(
      <Providers>
        <Volume volume={70} onVolumeChange={vi.fn()} />
      </Providers>
    );
    expect(container).toMatchSnapshot();
  });
});
