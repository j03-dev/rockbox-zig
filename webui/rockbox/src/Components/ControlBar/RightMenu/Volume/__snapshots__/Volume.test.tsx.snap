// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Volume > should render 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="css-1p6caij"
    >
      <div
        style="width: 100px; margin-top: 5px;"
      >
        <span
          class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeMedium css-m0hede-MuiSlider-root"
        >
          <span
            class="MuiSlider-rail css-r64h58-MuiSlider-rail"
          />
          <span
            class="MuiSlider-track css-xvk2i-MuiSlider-track"
            style="left: 0%; width: 70%;"
          />
          <span
            class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-b1410s-MuiSlider-thumb"
            data-index="0"
            style="left: 70%;"
          >
            <input
              aria-label="Volume"
              aria-orientation="horizontal"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="70"
              data-index="0"
              max="100"
              min="0"
              step="1"
              style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
              type="range"
              value="70"
            />
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class=""
  />
</div>
`;
